core/src/main/scala/org/apache/spark/status/StatusJsonHandler.scala

```diff
@@ -0,0 +1,168 @@
+
+    //TODO throw out all the JettyUtils stuff, so I can set the response status code, etc.
+    val servlet = new HttpServlet {
+      override def doGet(request: HttpServletRequest, response: HttpServletResponse) {
+        if (securityManager.checkUIViewPermissions(request.getRemoteUser)) {
+          response.setContentType("text/json;charset=utf-8")
+          route(request) match {
+            case Some(jsonRoute) =>
+              response.setHeader("Cache-Control", "no-cache, no-store, must-revalidate")
+              try {
+                val responseObj = jsonRoute.renderJson(request)
+                val result = jsonMapper.writeValueAsString(responseObj)
+                response.setStatus(HttpServletResponse.SC_OK)
+                response.getWriter.println(result)
+              } catch {
+                case iae: IllegalArgumentException =>
+                  response.setStatus(HttpServletResponse.SC_BAD_REQUEST)
+                  response.getOutputStream.print(iae.getMessage())
+              }
+            case None =>
+              println("no match for path: " + request.getPathInfo)
+              response.setStatus(HttpServletResponse.SC_NOT_FOUND)
+          }
+        } else {
+          response.setStatus(HttpServletResponse.SC_UNAUTHORIZED)
+          response.setHeader("Cache-Control", "no-cache, no-store, must-revalidate")
+          response.sendError(HttpServletResponse.SC_UNAUTHORIZED,
+            "User is not authorized to access this page.")
+        }
+      }
+    }
+    val path = "/json/v1"
+    val contextHandler = new ServletContextHandler
+    val holder = new ServletHolder(servlet)
+    contextHandler.setContextPath(path)
+    contextHandler.addServlet(holder, "/*")
+    contextHandler
+  }
+
+  def withSparkUI[T](request: HttpServletRequest)(f: (SparkUI, HttpServletRequest) => T): T = {
+    val appIdOpt = JsonRequestHandler.extractAppId(request.getPathInfo)
+    appIdOpt match {
+      case Some(appId) =>
+        uiRoot.getSparkUI(appId) match {
+          case Some(ui) => f(ui, request)
+          case None => throw new IllegalArgumentException("no such app: " + appId)
+        }
+      case None =>
+        throw new IllegalArgumentException("no app id")
+    }
+  }
+}
+
+private[spark] object JsonRequestHandler {
+  val appId = """/applications/([^/]+)""".r
+  val stageId = """/applications/[^/]+/stages/(\d+)""".r
+  val rddId = """/applications/[^/]+/storage/rdd/(\d+)""".r
+  def extractAppId(pathInfo: String): Option[String] = appId.findFirstMatchIn(pathInfo).map{_.group(1)}
+  def extractStageId(pathInfo: String): Option[Int] = stageId.findFirstMatchIn(pathInfo).map{_.group(1).toInt}
+  def extractRDDId(pathInfo: String): Option[Int] = rddId.findFirstMatchIn(pathInfo).map{_.group(1).toInt}
+}
+
+private[spark] trait UIRoot {
+  def getSparkUI(appKey: String): Option[SparkUI]
+  def getApplicationInfoList: Seq[ApplicationInfo]
+}
+
+
+object RouteUtils {
+  def extractParamSet[T <: Enum[T]](
```

Comment:
What does this method do?  I could read it and figure it out, but a one-sentence Scaladoc description would be helpful (maybe an example of an input and output)?